cpp/include/cudf/column/column.hpp

```diff
@@ -0,0 +1,255 @@
+  column(column&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column from existing device device memory.
+   *
+   * @note This constructor is primarily intended for use in column factory
+   * functions. 
+   *
+   * @param[in] dtype The element type
+   * @param[in] size The number of elements in the column
+   * @param[in] data The column's data
+   * @param[in] null_mask Optional, column's null value indicator bitmask. May
+   * be empty if `null_count` is 0 or `UNKNOWN_NULL_COUNT`.
+   * @param null_count Optional, the count of null elements. If unknown, specify
+   * `UNKNOWN_NULL_COUNT` to indicate that the null count should be computed on
+   * the first invocation of `null_count()`.
+   * @param children Optional, vector of child columns
+   *---------------------------------------------------------------------------**/
+  template <typename B1, typename B2 = rmm::device_buffer>
+  column(data_type dtype, size_type size, B1&& data, B2&& null_mask = {},
+         size_type null_count = UNKNOWN_NULL_COUNT,
+         std::vector<std::unique_ptr<column>>&& children = {})
+      : _type{dtype},
+        _size{size},
+        _data{std::forward<B1>(data)},
+        _null_mask{std::forward<B2>(null_mask)},
+        _null_count{null_count},
+        _children{std::move(children)} {}
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column by deep copying the contents of a
+   * `column_view`.
+   *
+   * This accounts for the `column_view`'s offset.
+   *
+   * @param view The view to copy
+   * @param stream The stream on which all allocations and copies will be
+   * executed
+   * @param mr The resource to use for all allocations
+   *---------------------------------------------------------------------------**/
+  explicit column(
+      column_view view, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the column's logical element type
+   *---------------------------------------------------------------------------**/
+  data_type type() const noexcept { return _type; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the number of elements
+   *---------------------------------------------------------------------------**/
+  size_type size() const noexcept { return _size; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the count of null elements.
+   *
+   * @note If the column was constructed with `UNKNOWN_NULL_COUNT`, or if at any
+   * point `set_null_count(UNKNOWN_NULL_COUNT)` was invoked, then the
+   * first invocation of `null_count()` will compute and store the count of null
```

Comment:
For this and other column methods that launch kernels underneath, should they should use the stream specified in the constructor or an optional stream argument?

Reply:
Hm, the fact that `null_count()` can potentially launch a kernel isn't something I want people to usually have to think about. That said, a `column` doesn't have a stream member, it's `device_buffer` members do, and I think it would be odd to use those. 

So I guess it makes sense to add a defaulted stream parameter to this function, but it makes me feel a bit uncomfortable since `null_count()` is supposed to be masquerading as just a simple `const` accessor.